packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -826,6 +826,20 @@ function voidFunctionArguments(
             thenableReturnIndices,
             voidReturnIndices,
           );
+          const contextualType = checker.getContextualTypeForArgumentAtIndex(
+            node,
+            index,
+          );
+          if (contextualType !== type) {
+            checkThenableOrVoidArgument(
+              checker,
+              node,
+              contextualType,
+              index,
+              thenableReturnIndices,
+              voidReturnIndices,
+            );
+          }
```

Comment:
[Bug] We should also do this both for arrays and tuples

```ts
declare function fn<T extends (...args: unknown[]) => unknown>(fn: T): void
fn<() => void>(async () => { }); // reported - ok

declare function tupleFn<T extends (...args: unknown[]) => unknown>(...fns: [T, string, T]): void
tupleFn<() => void>(async () => { }, 'foo', async () => { }); // not reported - bug

declare function arrayFn<T extends (...args: unknown[]) => unknown>(...fns: (T | string)[]): void
arrayFn<() => void>(async () => { }, 'foo', async () => { }); // not reported - bug
```

[playground link](https://deploy-preview-10042--typescript-eslint.netlify.app/play/#ts=5.5.2&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVNQB4AVeEADwxFWAGd4AKAOhdgHNaAueNAa1RwB3VAG0AugEp4AXgB8PVPyGpZDRKm7EJ3AG44swAFDrCDKXPh6DqqLQCe6RufkBveAF8JAbngB6X-BwAA44MFTA8AC08Di8hoagkLAIKOjYePAYyEEQIABiRKQUVDT0zKwwHNx8AsLizgpKwqosTOpc8CLEADTwtBgwWKhsvcSSuvpGWTn5RGYy8lbANvaO8xZu7r0A5Ig4ONu9tg5gTgvwm95%2BAQIYgSAhYSAR0QBGyGzxidBwSGiYuHwsBgUDsBRIZEo1DojFa7A6NWU9XOiOa5TaqA6DFIAB8%2BgMhmwJPUJgZDMDQeD1otJisTmcNh4dnsDkdVqdqRcPFd-PBbvdHuEovB3mwgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkRNHHRI4MAF8QKoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

Reply:
@auvred Thanks for the detailed examples. fixed 707d787